src/test/java/com/uid2/admin/salt/SaltRotationTest.java

```diff
@@ -483,4 +489,241 @@ void testRotateSaltsZeroWorksWhenThereIsFutureSaltFile() throws Exception {
         assertThat(result.getSnapshot().getExpires()).isEqualTo(daysLater(7).asInstant());
     }
+
+    @Test
+    void testRotateFromSaltToKey() throws Exception {
+        saltRotation = new SaltRotation(keyGenerator, JsonObject.of(AdminConst.ENABLE_V4_RAW_UID, true));
+        when(keyGenerator.generateRandomKeyString(anyInt())).thenReturn("random-key-string");
+
+        final Duration[] minAges = {
+                Duration.ofDays(30),
+        };
+
+        var willRefresh = targetDate();
+        var willNotRefresh = targetDate().plusDays(30);
+        var lastSnapshot = SaltSnapshotBuilder.start()
+                .entries(SaltBuilder.start().lastUpdated(targetDate().minusDays(60)).refreshFrom(willRefresh).currentSalt("salt1"),
+                        SaltBuilder.start().lastUpdated(targetDate().minusDays(60)).refreshFrom(willNotRefresh).currentSalt())
+                .build();
+
+        var result = saltRotation.rotateSalts(lastSnapshot, minAges, 1, targetDate());
+
+        assertThat(result.hasSnapshot()).isTrue();
+
+        var salt = result.getSnapshot().getAllRotatingSalts()[0];
+
+        assertThat(salt.currentSalt()).isNull();
+        assertThat(salt.previousSalt()).isEqualTo("salt1");
+        assertThat(salt.currentKey()).isNotNull();
+        assertThat(salt.currentKey().id()).isEqualTo(0);
+        assertThat(salt.currentKey().key()).isNotNull();
+        assertThat(salt.currentKey().salt()).isNotNull();
+        assertThat(salt.previousKey()).isNull();
+    }
+
+    @Test
+    void testKeyRotationInitialKeyIdPopulation() throws Exception {
+        saltRotation = new SaltRotation(keyGenerator, JsonObject.of(AdminConst.ENABLE_V4_RAW_UID, true));
+
+        final Duration[] minAges = {
+                Duration.ofDays(30),
+        };
+
+        var willRefresh = targetDate();
+        var willNotRefresh = targetDate().plusDays(30);
+        var lastSnapshot = SaltSnapshotBuilder.start()
+                .entries(SaltBuilder.start().lastUpdated(targetDate().minusDays(60)).refreshFrom(willRefresh).currentSalt(),
+                        SaltBuilder.start().lastUpdated(targetDate().minusDays(60)).refreshFrom(willRefresh).currentSalt(),
+                        SaltBuilder.start().lastUpdated(targetDate().minusDays(60)).refreshFrom(willRefresh).currentSalt(),
+                        SaltBuilder.start().lastUpdated(targetDate().minusDays(60)).refreshFrom(willNotRefresh).currentSalt())
+                .build();
+
+        var result = saltRotation.rotateSalts(lastSnapshot, minAges, 1, targetDate());
+        var salts = result.getSnapshot().getAllRotatingSalts();
+
+        assertEquals(0, salts[0].currentKey().id());
+        assertEquals(1, salts[1].currentKey().id());
+        assertEquals(2, salts[2].currentKey().id());
+        assertNull(salts[3].currentKey());
+    }
+
+    @Test
+    void testKeyRotationKeyIdIncrementation() throws Exception {
```

Comment:
Can we consolidate testKeyRotationKeyIdIncrementation, testKeyRotationKeyIdWrap, testKeyRotationKeyIdWrapManyBuckets?